File: apps/backend/app/api/routes/knowledge_bases.py
from typing import Annotated

from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.db.models.user import User
from app.db.session import get_db
from app.schemas.knowledge_base import KnowledgeBaseCreate, KnowledgeBaseRead, KnowledgeBaseUpdate
from app.services.knowledge_base_service import (
    create_knowledge_base,
    delete_knowledge_base,
    get_knowledge_base,
    list_knowledge_bases,
    update_knowledge_base,
)

router = APIRouter(prefix="/knowledge-bases")


@router.get("", response_model=list[KnowledgeBaseRead])
def list_items(
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[Session, Depends(get_db)],
) -> list[KnowledgeBaseRead]:
    return list_knowledge_bases(db, current_user.id)


@router.post("", response_model=KnowledgeBaseRead, status_code=status.HTTP_201_CREATED)
def create_item(
    payload: KnowledgeBaseCreate,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[Session, Depends(get_db)],
) -> KnowledgeBaseRead:
    return create_knowledge_base(db, current_user.id, payload)


@router.get("/{kb_id}", response_model=KnowledgeBaseRead)
def get_item(
    kb_id: str,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[Session, Depends(get_db)],
) -> KnowledgeBaseRead:
    return get_knowledge_base(db, current_user.id, kb_id)


@router.patch("/{kb_id}", response_model=KnowledgeBaseRead)
def update_item(
    kb_id: str,
    payload: KnowledgeBaseUpdate,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[Session, Depends(get_db)],
) -> KnowledgeBaseRead:
    return update_knowledge_base(db, current_user.id, kb_id, payload)


@router.delete("/{kb_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_item(
    kb_id: str,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[Session, Depends(get_db)],
) -> Response:
    delete_knowledge_base(db, current_user.id, kb_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: apps/frontend/src/lib/api/knowledgeBases.ts
import { apiRequest } from "./client";
import type { KnowledgeBase, KnowledgeBaseVisibility } from "./types";

export async function listKnowledgeBases(token: string): Promise<KnowledgeBase[]> {
  return apiRequest<KnowledgeBase[]>("/knowledge-bases", { token });
}

export async function createKnowledgeBase(
  token: string,
  payload: {
    name: string;
    description?: string;
    visibility?: KnowledgeBaseVisibility;
    department_id?: string | null;
  },
): Promise<KnowledgeBase> {
  return apiRequest<KnowledgeBase>("/knowledge-bases", {
    method: "POST",
    token,
    body: JSON.stringify({
      name: payload.name,
      description: payload.description ?? null,
      visibility: payload.visibility ?? "private",
      department_id: payload.department_id ?? null,
    }),
  });
}

export async function getKnowledgeBase(token: string, id: string): Promise<KnowledgeBase> {
  return apiRequest<KnowledgeBase>(`/knowledge-bases/${id}`, { token });
}

export async function updateKnowledgeBase(
  token: string,
  id: string,
  payload: {
    name?: string;
    description?: string | null;
    visibility?: KnowledgeBaseVisibility;
    department_id?: string | null;
  },
): Promise<KnowledgeBase> {
  return apiRequest<KnowledgeBase>(`/knowledge-bases/${id}`, {
    method: "PATCH",
    token,
    body: JSON.stringify(payload),
  });
}

export async function deleteKnowledgeBase(token: string, id: string): Promise<void> {
  await apiRequest(`/knowledge-bases/${id}`, { method: "DELETE", token });
}
